server/test/datadog.test.ts

```diff
@@ -36,19 +45,38 @@ describe("datadog middleware", () => {
     const res = await context.client.get<any>("api/edge/locations");
     expect(res.statusCode).toBe(200);
 
-    const middlewareMetrics = dogstatsd.mockBuffer.filter((s) =>
-      s.startsWith("node.express.router.")
-    );
-    expect(middlewareMetrics).toHaveLength(2);
+    if ("mockBuffer" in dogMetrics) {
+      // One of this stat, with a value of 1, and a correct response_code tag.
+      const responseCounts = dogMetrics.mockBuffer.filter((s) =>
+        s.startsWith("node.express.router.response_total:")
+      );
+      expect(responseCounts).toHaveLength(1);
+      expect(responseCounts[0]).toMatch(/^[\w.]+:1\b/);
+      expect(responseCounts[0]).toMatch(/\bresponse_code:200\b/);
 
-    expect(dogstatsd.mockBuffer).toContainEqual(
-      expect.stringMatching(
-        /^node\.express\.router\.response_total:1\|c\|#.*response_code:200.*/
-      )
-    );
-    expect(dogstatsd.mockBuffer).toContainEqual(
-      expect.stringMatching(/^node\.express\.router\.response_time:/)
-    );
+      // One of this stat.
+      const responseTimes = dogMetrics.mockBuffer.filter((s) =>
+        s.startsWith("node.express.router.response_time:")
+      );
+      expect(responseTimes).toHaveLength(1);
+    } else {
+      // @ts-expect-error They typings are missing `aggregator`.
+      const buffer: any[] = Object.values(dogMetrics.aggregator.buffer);
+      const responseCounts = buffer.filter(
+        (m) => m.key === "node.express.router.response_total"
+      );
+      expect(responseCounts).toHaveLength(1);
+      expect(responseCounts[0]).toHaveProperty("value", 1);
+      expect(responseCounts[0]).toHaveProperty(
+        "tags",
+        expect.arrayContaining(["response_code:200"])
+      );
+
+      const responseTimes = buffer.filter(
+        (m) => m.key === "node.express.router.response_time"
+      );
+      expect(responseTimes).toHaveLength(1);
+    }
```

Comment:
This mess is dumb. The right move would be to have test mocks on the `dogMetrics` object so we can just check which functions were called, but since the middle we’re testing and the `dogMetrics` object are in the same module, they are tightly coupled through local references and the middleware will never see the mock (because we can only really mock *between* modules).

I can resolve this by separating the middleware out of `datadot.ts`, but that would imply even more architectural changes, and it just didn’t feel worthwhile in *this* PR, which I’m viewing as more running an experiment than changing implementations.

Reply:
I was trying to use a Jest mock, but for some reason I was hung up on mocking the whole `dogMetrics` object as one (which you could only do by mocking the module, since it’s an export, and the problem is that if you mock the module, the middleware will still see the real thing instead of the mock). Coming back with fresh eyes, we could just mock each method on the `dogMetrics` object individually:

```ts
jest.spyOn(dogMetrics, "increment");
jest.spyOn(dogMetrics, "gauge");
jest.spyOn(dogMetrics, "histogram");

afterEach(() => {
  jest.clearAllMocks();
});

// ...tests...
```